### ISSUE TITLE
Cache Workers getting stuck?

### ISSUE THREAD

**Issue by normanuber** [ben-manes/caffeine#127]
I saw this problem in version 2.3.1, via VisualVm, when after a few hours that a bunch of ForkJoinWorkers would be stuck on the WriteBuffer.poll() function in RUNNABLE state. Even hit an OOM when the cpu utilization became too high and my events started to get backed up. I recently upgraded to 2.3.3 and thought the problem was fixed until the below screenshots.

When I started up the application (10/19, 7:34 am) there is one FJW in the writebuffer.poll, which is fine.
<img width="998" alt="screen shot 2016-10-19 at 7 34 41 am" src="https://cloud.githubusercontent.com/assets/14983089/19569558/0a6ad1bc-96ab-11e6-9706-3505dd75cae3.png">

This morning (10/20, 9:34am) and this is in constant RUNNABLE state. The number of workers that will be in this stuck runnable state will slowly grow over time too because I have seen this behavior happen with 2.3.1.
<img width="993" alt="screen shot 2016-10-20 at 9 34 02 am" src="https://cloud.githubusercontent.com/assets/14983089/19569552/01d2bcfe-96ab-11e6-8d04-0b0936cf43a9.png">

The cpu utilization and cpu load metrics in Grafana support an overall increase work as if the workers are busy waiting/spinlooping or something.

There are maybe 1k or so instances of the cache data structure in the application. This server has the most instances of the cache and the issue appears to manifest faster than the other servers. All of the servers have a 64GB heap.

I will have to switch back to Guava considering this is a production application.

Thanks for your effort though!

**Comment by ben-manes** [comment]
Thanks for the detailed report. I'm sorry for the inconvenience this caused.

The previous challenges with the write buffer were around scheduling the task when it is being added. Some of that was due to a FJP bug, and then later mistakes when I tried to make the logic more robust. Despite a lot of effort put into testing, its hard to predict some of these concurrency bugs.

This case is different because only 1 thread (for a given cache) can poll the write buffer. This drains and replays the changes onto the eviction policy. That polls up to a maximum, so the spin loop causing the failure is inside the `WriteBuffer`. That's a stripped down version of JCTools' [MpscChunkedArrayQueue](https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscChunkedArrayQueue.java).

I think the first step is try to reproduce the live lock, then re-sync with the latest JCTools version, and if broken then work with @nitsanw on a fix.

**Comment by normanuber** [comment]
Thanks for acknowledging the issue! I hope you can reproduce it. If you need any more info, I'll try my best to help.

**Comment by nitsanw** [comment]
@ben-manes "I think that I can weaken JCTool's spin loop safely, because the element will be consumed on the next maintenance cycle. That is scheduled after the write buffer is appended to. So assuming the buffer isn't losing the element, it should be visible the next time around." - this approach will not work. If producerIndex is not consumerIndex, and the consumer is not making progress then they cannot be equal on the next cycle, which brings you back to looking for the next element in the array, which is not there because if it was you'd not be stuck in that loop.

@normanuber Can you please quote the JVM version (java -version output) and OS? it's a long shot, but certain JVM builds have concurrency bugs and it's a variable that is easy to eliminate.

At a glance the code seems very similar to the original, I would have to diff line by line to verify. Is there absolutely no way that 2 different threads are acting as consumers?

**Comment by nitsanw** [comment]
@normanuber would you be able to run a hacked caffeine build to help narrow down the source of this issue?

**Comment by ben-manes** [comment]
The polling occurs in [drainWriteBuffer()](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1133) which is guarded by `evictionLock`. In addition to my own tests, static analysis performed by ErrorProne should be asserting this behavior. So only a single thread should be able to consume the write buffer.

**Comment by normanuber** [comment]
@nitsanw 

java -version
OpenJDK Runtime Environment (build 1.8.0_102-8u102-b14.1-1~bpo8+1-b14)
OpenJDK 64-Bit Server VM (build 25.102-b14, mixed mode)

lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 8.6 (jessie)
Release:	8.6
Codename:	jessie

To answer your question, I suppose I can run a hacked build as long as it is through maven central (using SNAPSHOT versions is fine).

**Comment by ben-manes** [comment]
I created a SNAPSHOT version that adds jctools as a dependency and uses `MpscChunkedArrayQueue` directly. That should avoid any surprises if I incorrectly embedded the code. See the [jctools branch](https://github.com/ben-manes/caffeine/tree/jctools) for this commit.

While that is running on travis, I've manually pushed a new SNAPSHOT. The travis build will take about an hour to verify against the tests before releasing a snapshot.

**Comment by nitsanw** [comment]
@ben-manes Thanks!

**Comment by ben-manes** [comment]
@normanuber Since you are using a high number of caches (1,000) it might be good to quickly discuss the use-case. For example if these are per-request, quickly populated and discarded, it could lead to a pile-up of cache instances and async processing tasks scheduled. The concurrency and hit-rate focus of this cache leads to a design towards long lasting instances with high utilization. It may be that this is mitigated using `Caffeine.executor(Runnable::run)` to penalize the caller or, if low concurrency, a `LinkedHashMap` in LRU mode is a better match.

It may be that we need to better handle high instance churn, which is observable as a filling up FJP with stale write tasks and we'd mistakenly focus trying to break the queue.

**Comment by nitsanw** [comment]
@normanuber please report back if the issue is still present with vanilla JCTools 1.2.1
I'm holding off a release on the chance that this is an issue I need to resolve.
Thanks!

**Comment by normanuber** [comment]
Looks like the email didn't make it here.

I've already switched my production application back to Guava since I require removal notifications or else I would use a CHM.

There are different levels of the cache, but the most threads touching any of the caches is below 10. 4 dedicated writers for the highest level cache while less than 6 readers at any time reading from it. The highest level cache is a permanent data structure while the child data structures containing a cache last as long as the child element will last (these children can last for weeks). A few hundred of these datastructures are expired/created per day from what my stats say.

I will setup a test environment around the application and attempt to include your library when github is back up.

**Comment by normanuber** [comment]
Is the snapshot version in Maven Central? If not can you push one out there?

**Comment by ben-manes** [comment]
So far I have been unable to construct a test case that breaks `WriteBuffer`. It could be a OS/JVM/hardware match-up causing a visibility problem.

My best alternative guess is that you're workload is causing the cache to flood FJP. That's used to avoid penalizing the caller for policy maintenance, e.g. writes triggering eviction, and removal notifications. If you configure with `Caffeine.executor(Runnable::run)` then it will behave similarly to Guava by amortizing the cost on the callers. The only difference is that Guava uses a notification queue to process outside of any lock, whereas with an executor we assume it can be delegated to inside the policy's lock. So unless your removal listener is very expensive this won't make a big difference.

---

Maven Central does not allow snapshots and has a dedicated repository for that workflow instead. You need to add that repository to your pom as,

```xml
<repositories>
  <repository>
    <id>snapshots</id>
    <url>http://oss.sonatype.org/content/repositories/snapshots</url>
    <releases>
      <enabled>false</enabled>
    </releases>
    <snapshots>
      <enabled>true</enabled>
    </snapshots>
  </repository>
</repositories>
```

**Comment by normanuber** [comment]
I'm am attempting to repro the issue on the test environment with 2.3.1 then 2.3.3 finally 2.3.4-SNAPSHOT and will get back to you.

**Comment by ben-manes** [comment]
Thanks @normanuber. From https://github.com/JCTools/JCTools/issues/135 we might want to try using the latest snapshot of JCTools' to see if that change fixes the problem.

**Comment by normanuber** [comment]
Alrighty, so all of them got stuck (2.31 - 2.34-SNAPSHOT). I tested in production instead >_>, since the test environment is too small.

2.31 was the fastest one to get stuck. Maybe 1-2 mins to get a stuck worker.
<img width="983" alt="screen shot 2016-10-23 at 1 19 31 pm" src="https://cloud.githubusercontent.com/assets/14983089/19629405/6f804d18-9929-11e6-9de0-ccc1199ad6f1.png">

2.33 at about 12 minutes in
<img width="996" alt="screen shot 2016-10-23 at 1 33 10 pm" src="https://cloud.githubusercontent.com/assets/14983089/19629402/6f7d2c64-9929-11e6-9912-8e0457817312.png">
<img width="988" alt="screen shot 2016-10-23 at 1 34 08 pm" src="https://cloud.githubusercontent.com/assets/14983089/19629403/6f7d347a-9929-11e6-88bd-7a3529cf6997.png">

2.34-SNAPSHOT about 27- 30 mins
<img width="995" alt="screen shot 2016-10-23 at 2 00 12 pm" src="https://cloud.githubusercontent.com/assets/14983089/19629404/6f7f6d9e-9929-11e6-91a4-a71831b055f5.png">
<img width="994" alt="screen shot 2016-10-23 at 2 00 26 pm" src="https://cloud.githubusercontent.com/assets/14983089/19629401/6f7d1e90-9929-11e6-9213-0709e50d2616.png">

2.34 FJP
<img width="871" alt="screen shot 2016-10-23 at 2 05 29 pm" src="https://cloud.githubusercontent.com/assets/14983089/19629414/d5795632-9929-11e6-8394-c0e678d28629.png">

**Comment by ben-manes** [comment]
Can you test if I update to JCTools' snapshot?

**Comment by normanuber** [comment]
Sure, tell me when you have updated.

**Comment by ben-manes** [comment]
I uploaded a new snapshot. However since Maven only checks 24-hours you may need to clean force it to update (`mvn clean install -U`). If that's too much trouble, you can specify the snapshot timestamp explicitly as `2.3.4-20161023.211902-21`.

JCTools does not deploy snapshots and leverages jitpack.io instead. You'll need to add that repository as well.
```xml
<repository>
    <id>jitpack.io</id>
    <url>https://jitpack.io</url>
  <releases>
    <enabled>true</enabled>
  </releases>
  <snapshots>
    <enabled>true</enabled>
  </snapshots>
</repository>
```

**Comment by normanuber** [comment]
1 hour runtime so far and no problems. I'll keep it running for 4 more hours before rolling back to the stable version on my software.

**Comment by normanuber** [comment]
Loos good after 5 hours of runtime.

**Comment by ben-manes** [comment]
Thanks @normanuber and @nitsanw!

@nitsanw do you think its possible to write a unit test that exposes this bug and its fix?

Since the next version of JCTools is imminent, I'd like to port over tied to a release. But if that gets delayed I'll try to have one out soon regardless. I'll revisit whether to shadow or add a dependency later.

When this fix is released I'll close this issue. Thanks again for the help and patience.

**Comment by nitsanw** [comment]
Thanks @normanuber for your efforts in testing and verifying :-)
Thanks @ben-manes, I'll be in touch but expect a release this week. I would love for you to shadow JCTools, but I understand it can't be your top priority :-)

**Comment by johnou** [comment]
@nitsanw unfortunately the gradle shadow library doesn't support minimizeJar https://github.com/johnrengelman/shadow/issues/115

**Comment by ben-manes** [comment]
It also doesn't support changing the class visibility, so it becomes part of the public API. If a direct dependency, then JCTools would be expected to have a stronger (semver) compatiblity story. E.g. the constructor for the queue was removed in 1.3 in favor of a new subclass. For now I prefer manually shading but jigsaw might remove my concerns, so it's good to keep reconsidering the current approach.

**Comment by johnou** [comment]
@ben-manes you could also relocate JCTools to an "internal" package, nice warning for developers.

**Comment by ben-manes** [comment]
I know, but it's really annoying as IDE automatic imports often bring in the wrong one. So developers get frustrated by the conflicts of internal versions of common classes. I'm hoping jigsaw alleviates that mess.

**Comment by johnou** [comment]
Hahaha several years ago when auto-completion was taking off I ran into some really exotic imports in code reviews :)

**Comment by johnou** [comment]
@normanuber awesome work!

**Comment by nitsanw** [comment]
@ben-manes "JCTools would be expected to have a stronger (semver) compatiblity story. E.g. the constructor for the queue was removed in 1.3 in favor of a new subclass."
1.3 is not out, and given the above should perhaps be called 2.0. Thanks for the feedback.

**Comment by normanuber** [comment]
I've already switched my production application back to Guava since I
require removal notifications or else I would use a CHM.

There are different levels of the cache, but the most threads touching any
of the caches is below 10. 4 dedicated writers for the highest level cache
while less than 6 readers at any time reading from it. The highest level
cache is a permanent data structure while the child data structures
containing a cache last as long as the child element will last (these
children can last for weeks). A few hundred of these datastructures are
expired/created per day from what my stats say.

I will setup a test environment around the application and attempt to
include your library when github is back up.

**Comment by nitsanw** [comment]
@ben-manes note that v2.0 is out with fix. I can do a 1.2.2 bugfix release if it helps your cause.

**Comment by ben-manes** [comment]
Thanks @nitsanw, I really appreciate all of the hard work you put into jctools.

I'll have a release out in the day or two, Sunday at the latest.

**Comment by ben-manes** [comment]
@nitsanw Can you please publish the test artifacts in future releases? Since I'm manually shading, I'd like to copy and run `MpscGrowableQueueSanityTest` against it. Rather than copy `QueueSanityTest` and its imports, simply having the test dependency would be ideal. Not sure how to query jitpack for this type of artifact.

**Comment by ben-manes** [comment]
Travis will take 1.5hrs to complete. Once the build passes I'll cut a release.

**Comment by ben-manes** [comment]
Released `2.3.4`. I'm grateful for everyone's help, speedy responses, and patience.
